Dev

sssion opened the pull request.

sssion committed:
update the md

sssion committed:
create fundamental structures and linksystem, schema...

sssion committed:
add base node functions and cursor functions, create ProllyTree struct with search, get api

sssion committed:
finish part of tree build procedure, add weibullSplitter

sssion committed:
finish tree build procedure

sssion committed:
add some functions to optimize procedure of tree build

sssion committed:
add go mod

sssion committed:
1. replace level(int) in node with IsLeaf(boolean)
2. delete links in node, transform value by the IsLeaf

sssion committed:
1. add ProllyRoot
2. remove configCid in ProllyNode

sssion committed:
fix some wrong comments

sssion committed:
add test codes

sssion committed:
implement suffix splitter and modify the schema, clean code

sssion committed:
fix bugs and add put and rebalance procedure

sssion committed:
optimize the procedure of tree build

sssion committed:
1.update schema, add nodeCodec(prefix) in config
2.add NodeCoder struct
3.prefix.codec is used in Size computing and prefix is used in node saving
4. rename the PrefixThreshold function to SuffixThreshold function

sssion committed:
add batch option

sssion committed:
fix bugs in tree mutating and building.
Add unittests

sssion committed:
init the prollytree adl framework

sssion committed:
finish base workflow for adl

sssion committed:
add part of adl implementation

sssion committed:
redesign the project struct, fix bug of lru pointer pollution and add ProllyTreeNode into ProllyTree

sssion committed:
clean code and rename some struct

sssion committed:
clean code, rename some files, replace some pointer by struct to avoid pollution

sssion committed:
add MapIterator() for adl node, update the search function in prollytree for adding ending key

sssion committed:
update the mutations, only sort the mutations pairs after finished

sssion committed:
add fixtures and unittest, rename some structs and clean code

taoshengshi committed:
add draft documents

sssion committed:
add codec schema and struct in config

taoshengshi committed:
add prolly trees introduction

sssion committed:
fix bug in tree mutations and modify part workflow of tree rebuild

sssion committed:
replace cidprefix bytes with nodecodec,  clean code and rename some files

sssion committed:
add special fixture

sssion committed:
update fixtures data type

RangerMauve commented on pkg/tree/iterator.go:
```diff
@@ -0,0 +1,59 @@
+	}
+}
+
+type Iterator struct {
```
Does golang have any standards for iterators we can work off of? What does the rest of IPLD use?

Would it make to use a channel here or expose the data as a channel?

sssion replied:
The iterator implements the MapIterator( https://github.com/ipld/go-ipld-prime/blob/1a409dc236dd1697fa787ca7166e43c884a57d0e/datamodel/node.go#L259 )interfaces in ipld node(map kind), so I reuse it in the search api and adl. Should we have more methods to support the iterator?

RangerMauve commented on pkg/tree/schema.ipldsch:
```diff
@@ -0,0 +1,41 @@
+    RootCid &ProllyNode
+    # config about the tree
+    ConfigCid &TreeConfig
+}
```
Should also be representation tuple

RangerMauve commented on pkg/tree/schema.ipldsch:
```diff
@@ -0,0 +1,41 @@
+
+type SuffixThresholdConfig struct {
+    chunkingFactor Int
+} representation tuple
```
Should also have hashFunction int https://github.com/ipld/ipld/pull/254/files#diff-b6100ac62d9896fb718b68b3208c91a1ba120b46253b7d95e0a65d4afaf72fe8R97

RangerMauve commented on pkg/tree/schema.ipldsch:
```diff
@@ -0,0 +1,41 @@
+    maxPairsInNode  Int
+    nodeCodec       Prefix
+    strategy        strategy
+} representation tuple
```
Please use the same ordering of keys as in the spec code. MaxSize, MinSize, cidVersion, codec, hashFunction, nullable hashLength, strategy ChunkingStrategy

sssion replied:
maybe we should strategyType(Int) to specify the strategy type? It makes code easy to get the specific strategy, because the strategy in Go is like below, we will check every strategy without type 
type strategy struct {
	//Weilbull    *WeibullThresholdConfig
	//RollingHash *RollingHashConfig
	Suffix *PrefixThresholdConfig
}

RangerMauve replied:
Ah, does Go not have a way to say "This field is one of three types"?

sssion replied:
As I know, there isn't convenient way to get the real config from the union struct in Go

RangerMauve replied:
We should get rid of the union and split it into multiple keys in the spec, then.

RangerMauve commented on pkg/tree/schema.ipldsch:
```diff
@@ -0,0 +1,41 @@
+} representation tuple
+
+type strategy union {
+    | SuffixThresholdConfig     "suffix"
```
Should use same naming as in the spec

RangerMauve commented on pkg/tree/schema.ipldsch:
```diff
@@ -0,0 +1,41 @@
+    | SuffixThresholdConfig     "suffix"
+} representation keyed
+
+type Prefix struct{
```
Where is this being used?

sssion replied:
I extract the fields below in ProllyTreeConfig as the Prefix struct, they are used in TreeNode saving(in NodeStore and get cid), we should put these fields in ProllyTreeConfig directly?
  cidVersion Int
  codec Int
  hashFunction Int
  nullable hashLength Int

RangerMauve replied:
In the spec we had it in the config itself. I think it'd be useful to have it all in one place

RangerMauve commented on pkg/tree/adl_assembler.go:
```diff
@@ -0,0 +1,188 @@
+package tree
```
Maybe we can put the ADL stuff into its own folder to make it easier to navigate?

sssion replied:
yes, I will do this,  and I find that ipld node serializing need n.Length() (at least dagcbor for map node),  so maybe we need add subtreesize( subtree pairs count) field in schema?
<img width="962" alt="image" src="https://user-images.githubusercontent.com/62473385/213413491-708e2f5e-aa37-49e2-946a-3c3370c59c19.png">

RangerMauve replied:
Oh I see. 🤔 Does it error out if you give it too large a length? @warpfork any thoughts on what we should do about this? Adding a field to track subtree size seems useful since UnixFS does similar, but I wonder if there are any alternatives we could consider.

RangerMauve reviewed (commented):
This looks good so far. I think it'd be important to make sure the schema is exactly the same as in the spec so that all implementations are compatible.

I'll be doing a more thorough inspection of the construction algorithm later but at first glance it looks pretty close to the spec text. 👍

It would be useful if we could run a flame graph along side some benchmarks to get an idea of where our performance bottlenecks are.

https://www.matoski.com/article/golang-profiling-flamegraphs/

sssion commented on pkg/tree/framework.go:
```diff
@@ -0,0 +1,566 @@
+		}
+		ok = lb.nodeBuffer.tryAddPair(key, value)
+		if !ok {
+			panic("too large pair bigger than the node size limit")
```
If code arrives the panic, it means that after splitting and reset, the single pair is over the MaxNodeSize(the node saves only one pair is still over size)

sssion replied:
If user inputs a large key/value pair(eg. over 4kb) may lead to this cause

sssion committed:
1.move adl files to new directory and modify the adl node struct
2.rename some schema fields
3.add hashFunction field in HashThresholdConfig and update the splitter

sssion committed:
add subtreeCount field in ProllyNode and implement adl node Length() function, modify the type of the key got from MapIterator from Bytes(ipld kind) to String(ipld kind)

sssion committed:
commit fixtures files

sssion committed:
merge NodeCodec into TreeConfig and clean code

burdiyan commented on go.mod:
```diff
@@ -0,0 +1,66 @@
+module go-ipld-prolly-trees
```
Please, consider naming the module after the GitHub URL to make it go-getable.

RangerMauve replied:
Agreed, would nbe good to check that we can add this module easily via go-get.

RangerMauve replied:
Fixed. Use `go env -w GOPROXY="direct"` if you have trouble installing the latest commit.

burdiyan commented on pkg/tree/types.go:
```diff
@@ -0,0 +1,34 @@
+		Version:  1,
+		Codec:    uint64(multicodec.DagCbor),
+		MhType:   uint64(multicodec.Sha2_256),
+		MhLength: 16,
```
Why setting digest length to 16 bytes instead of the default 32? Seems to be pretty uncommon in the IPFS ecosystem.

RangerMauve replied:
I think to save on overall bytes and to fit more CIDs/KVs into a node.

Is the risk of collision too high with 16?

RangerMauve reviewed (approved):
Good to go, will update the schema in the spec

sssion committed:
update the DefaultLinkProto(the MhLength) to make the car can be loaded by `ipfs dag import`

Commit:
rename package to be go gettable

Commit:
propogate module rename

Commit:
Allow building empty prolly trees with framework

sssion committed:
replace the search iterator with channel(array before), now it works async. And delete the mutex in iterator.

Commit:
Add fialing test for doing a put into empty trees

Commit:
Expose TreeConfig from prolly tree

Commit:
Add flag to see if tree is being mutated

Commit:
Run fmt

Commit:
Add GetProof method

Commit:
Fix returning proof for leaf nodes

sssion committed:
fix bug for empty tree mutating and wrong processing about adding key bigger than all keys in the tree
add ut for the critical case

Commit:
Rewrite GetProof to include indexes

Commit:
Make proof segment properties public

sssion committed:
add skipCommon function

Commit:
Track tree cid in tree for use in getproof

Commit:
Set treeCid when loading from nodestore

Commit:
Add more treeCid and getproof tests, fix missing cid

Commit:
Add missing root tree node in proof

sssion committed:
add diff and merge api to support ProllyTree merge option, and add some uts

sssion committed:
fix bug in skipCommon and add some uts for tree merge function

sssion committed:
add ut for merge tree

sssion committed:
fix bug

sssion committed:
add Proof and ProofSegment ipld schema

sssion committed:
add Proof write/read function in NodeStore interface

sssion committed:
add doc

sssion commented:
@RangerMauve Hi, I add a spec doc in the projcet docs dir, could you please have a reivew?

taoshengshi commented:
> @RangerMauve Hi, I add a spec doc in the projcet docs dir, could you please have a reivew?

@sssion Very good  illustration of implementation! Could you add these good content into markdown files other than docx files?

sssion committed:
add the markfown file of spec

RangerMauve commented on docs/IPLD-ProllyTree-Spec.md:
```diff
@@ -0,0 +1,271 @@
+# IPLD Prolly Tree Spec
+## Summary
+IPLD Prolly Tree is a index module that is based on the ProllyTree and IPLD.It uses the ProllyTree algorithm(combination of B-Tree and Merkle Tree) but the storage and tree nodes are implemented by IPLD node and store.So IPLD ProllyTree has advantages that exit in IPLD and ProllyTree such as any IPLD node input, structure share,  low updating while tree mutating and so on.And it can be used as so ...
```
```suggestion
IPLD Prolly Tree is a index module that is based on the ProllyTree and IPLD. It uses the ProllyTree algorithm(combination of B-Tree and Merkle Tree) but the storage and tree nodes are implemented using the IPLD suite of tools and specifications. IPLD ProllyTrees have advantages from both IPLD and ProllyTree such as supporting any IPLD node for values, deduplication, reducing the number of changed nodes when updating a tree, and more. It can be used as a database index similar to Dolt (via our IPLD-Prolly-Indexer library), or used as an ADL. It’s powerful in some fields like fact diff, version control, It works really well for storing structured data with git-like semantics.
```

RangerMauve commented on docs/IPLD-ProllyTree-Spec.md:
```diff
@@ -0,0 +1,271 @@
+
+![img8](./images/spec/img_8.png)
+
+IPLD ProllyTree implemented the IPLD ADL(Advanced Layout) interface based on above structures.It includes ProllyTreeADLPrototype, Builder, TreeAssembler(MapAssembler) and ADL Node. It helps people who want to build large map(dict) by ProllyTree with ADL in IPLD world.
```
```suggestion
IPLD ProllyTree implemented the IPLD ADL(Advanced Data Layout) interface based on above structures. It includes ProllyTreeADLPrototype, Builder, TreeAssembler(MapAssembler) and ADL Node. It helps people who want to build large map(dict) by ProllyTree with ADL in IPLD world.
```

RangerMauve commented on docs/IPLD-ProllyTree-Spec.md:
````diff
@@ -0,0 +1,271 @@
+}
+```
+
+NodeStore is an import module that saves almost all information and data of ProllyTree. It is used anywhere in the project such as tree building, updating, getting…… We can also read the saved tree before. The nodestore we use in the project is based on the blockstore(ipfs)
````
```suggestion
NodeStore is an important module that saves almost all information and data of ProllyTree. It is used anywhere in the project such as tree building, updating, getting, etc. It can also read previously saved tree nodes. The nodestore we use in the project is based on the IPFS blockstore interface.
```

RangerMauve commented on docs/IPLD-ProllyTree-Spec.md:
````diff
@@ -0,0 +1,271 @@
+mutations *Mutations
+}
+```
+ProllyTree is the core part in the project, it provides api for people to use the tree, it can also play the role as a package like in ADL, IPLD-Prolly-Indexer. The root node(ProllyNode) is the entry of the tree. When we need traverse the tree or some branch path in tree, we read the cidlink in node value and get relative node from NodeStore by cid.TreeConfig contains important information about  ...
````
```suggestion
ProllyTree is the core part in the project, it provides api for people to use the tree, it can also play the role as a package like in ADL, IPLD-Prolly-Indexer. The root node(ProllyNode) is the entry of the tree. When we need traverse the tree or some branch path in tree, we read the cidlink in node value and get relative node from NodeStore by cid.TreeConfig contains important information about the prollytree such as node size limit, hash function, cid version, codec and so on.It is set while building from framework.Mutations is used for tree mutating(updating), we won’t update for every mutating action such as putting(includes inserting and modifying), deleting. Mutations caches these actions then apply them when calling `Rebuild()`.
```

RangerMauve reviewed (changes requested):
excellent work! added some small suggestions for wording

I think with the markdown document version it might be best to remove the microsoft word file to save on space and be more compatible with what folks expect on Github.

also do you think you could do a once over and make sure they're spaces after each. and new lines after each header?

sssion committed:
Update docs/IPLD-ProllyTree-Spec.md

Co-authored-by: RangerMauve <RangerMauve@hotmail.com>

sssion committed:
Update docs/IPLD-ProllyTree-Spec.md

Co-authored-by: RangerMauve <RangerMauve@hotmail.com>

sssion committed:
Update docs/IPLD-ProllyTree-Spec.md

Co-authored-by: RangerMauve <RangerMauve@hotmail.com>

sssion committed:
Update docs/IPLD-ProllyTree-Spec.md

Co-authored-by: RangerMauve <RangerMauve@hotmail.com>

sssion committed:
delete doc file of spec and reformat the spec

sssion committed:
delete redundant words

sssion committed:
format the spec

sssion commented:
> excellent work! added some small suggestions for wording
> 
> I think with the markdown document version it might be best to remove the microsoft word file to save on space and be more compatible with what folks expect on Github.
> 
> also do you think you could do a once over and make sure they're spaces after each. and new lines after each header?

Thanks for your comment, I have updated the spec.

taoshengshi reviewed (approved):
Great work!
